src/constructUrl.js

```diff
@@ -113,49 +114,176 @@ var DEFAULT_OPTIONS = Object.freeze({
-  }
-
-  const keys = Object.keys(longOptions);
-  const keysLength = keys.length;
-  let url = src + "?";
-  for (let i = 0; i < keysLength; i++) {
-    let key = keys[i];
-    let val = longOptions[key];
+  const params = compactParamKeys(longOptions);
+  const { client, pathComponents } = extractClientAndPathComponents(src);
+  return client.buildURL(pathComponents.join("/"), params);
+}
 
+function compactParamKeys(longOptions, width, height) {
+  const params = Object.keys(longOptions).reduce((acc, key) => {
     if (PARAM_EXPANSION[key]) {
-      key = PARAM_EXPANSION[key];
+      acc[PARAM_EXPANSION[key]] = longOptions[key];
     } else {
-      key = encodeURIComponent(key);
+      acc[key] = longOptions[key];
     }
+    return acc;
+  }, {});
 
-    if (key.substr(-2) === "64") {
-      val = Base64.encodeURI(val);
-    }
+  if (width) {
+    params["w"] = width;
+  }
 
-    url += key + "=" + encodeURIComponent(val) + "&";
+  if (height) {
+    params["h"] = height;
   }
+  return params;
+}
+
+function extractClientAndPathComponents(src) {
+  const [scheme, rest] = src.split("://");
+  const [domain, ...pathComponents] = rest.split("/");
+  let useHTTPS = scheme == "https";
+
+  const client = new ImgixClient({
+    domain: domain,
+    useHTTPS: useHTTPS,
+    includeLibraryParam: false,
+  });
 
-  return url.slice(0, -1);
+  return { client, pathComponents };
 }
 
 function buildURLPublic(src, imgixParams = {}, options = {}) {
   const { disableLibraryParam } = options;
 
   const [rawSrc, params] = extractQueryParams(src);
 
-  return constructUrl(
-    rawSrc,
-    Object.assign(
-      {},
-      params,
-      imgixParams,
-      disableLibraryParam ? {} : { ixlib: `react-${PACKAGE_VERSION}` }
-    )
-  );
+  return constructUrl(rawSrc, {
+    ...params,
+    ...imgixParams,
+    ...(disableLibraryParam ? {} : { ixlib: `react-${PACKAGE_VERSION}` }),
+  });
+}
+
+/**
+ * Build a imgix source url with parameters from a raw url
+ */
+function buildSrc({
+  src: inputSrc,
+  width,
+  height,
+  disableLibraryParam,
+  disableSrcSet,
+  imgixParams,
+  disableQualityByDPR,
+}) {
+  const [rawSrc, params] = extractQueryParams(inputSrc);
+  const fixedSize = width != null || height != null;
+
+  const srcOptions = {
+    ...params,
+    ...imgixParams,
+    ...(disableLibraryParam ? {} : { ixlib: `react-${PACKAGE_VERSION}` }),
+    ...(fixedSize && height ? { height } : {}),
+    ...(fixedSize && width ? { width } : {}),
+  };
+
+  if (disableSrcSet) {
+    const src = constructUrl(rawSrc, srcOptions);
+    return { src, src };
+  }
+
+  if (fixedSize) {
+    const srcSet = buildSrcSet(
+      rawSrc,
+      srcOptions,
+      {
+        disableVariableQuality: disableQualityByDPR,
+      },
+      width,
+      height
+    );
+
+    const src = constructUrl(rawSrc, srcOptions);
+
+    return { src, srcSet };
+  } else {
+    warnInvalidAspectRatio(imgixParams.ar || "", config);
+
+    const { w, h, ...urlParams } = srcOptions;
```

Comment:
It seems like this destructuring is unnecessary since this behaviour is handled by `fixedSize`?

Reply:
Yeah, that's what I thought too. However, if you remove `w` and `h` you reveal a subtle bug in the implementation logic. Before these changes I think the original impl was destructuring to remove `w` and `h` from the options and only propagate `urlParams` down through the rest of the logic.

I think there may need to be some three-way-merging needed here? Something like "combine/resolve which height and width to use given those from props, ixParams, and from extracting from a src."